### PR TITLE
RavenDB-17312 Using different approach to ensure the same order of processing group by fields during calculation of the reduce key hash

### PR DESCRIPTION
**Pull request by arekpalinski**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17312/PropertyAccessor.GetPropertiesInOrder-does-not-preserve-order

### Additional description

We're not able to tell the property accessor what is the expected order because in JS indexes we're not always able to extract field names. At the same time field options can be specified so we know about _some_ fields. This caused indexing errors reproduced by the added test.

Now instead of relaying on the order of all properties returned by the property accessor we'll ensure the order of processing only group by fields at higher level - inside the map-reduce index.

This reverts changes made in https://github.com/ravendb/ravendb/pull/15413 and uses different approach to address a possible reduce key hash calculation inconsistency.

### Type of change

- Bug fix

### How risky is the change?
 
- Moderate 

### Backward compatibility

- Ensured. Only new indexes will use this approach - the internal index version was incremented

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?
- No

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs [hunk 1]
```diff
@@ -151,11 +151,11 @@ public static class IndexVersion
 
             public const long EngineTypeStored = 54_000; // introducing Corax, added engine type to the index storage
 
-            public const long GuaranteedOrderOfPropertiesInMapReduceIndexes = 54_001; // RavenDB-17312
+            public const long GuaranteedOrderOfGroupByFieldsInMapReduceIndexes = 54_002; // RavenDB-17312 - version 54_001 had an issue so we had to fix it and bump version again
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs:156]: What will happen when the index is built on `54_001`?  Is it safe to switch them to the `legacy` version of the order?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs:156]: You, are right. I need to have the code using version `54_001` as well

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs:156]: That is a bit tricky. In version `54_001` we ensured the same order of _all_ properties. Now we sort only group by fields. This means that the order of group by fields remained the same. So we can use the same code. But previously we sorted fields  _only_ if we had any map fields (in JS indexes we can have none):
https://github.com/ravendb/ravendb/blob/6a5cabb0d13fb576ca99d98ec9475519d36d59f9/src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs#L490-L493

This means we still need to recognize this case. In https://github.com/ravendb/ravendb/pull/15822/commits/86fc1734e040def6a37a857d7e6fb3d01db4c7c9 I ensured the backward compatibility and added the test.

**Review by ppekrol — APPROVED** [review]

**Comment by arekpalinski** [comment]
test this please

**Review by maciejaszyk — APPROVED** [review]

**Pull request merged**